core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/conditional/CaseExpression.kt

```diff
@@ -0,0 +1,159 @@
+            queryString = caseDopeQuery.queryString +
+                whenThenDopeQueries.joinToString(" ", " ", " ") { "$WHEN ${it.first.queryString} $THEN ${it.second.queryString}" } +
+                END,
+            parameters = caseDopeQuery.parameters + whenThenDopeQueries.fold(emptyMap()) { parameters, query ->
+                parameters + query.first.parameters + query.second.parameters
+            },
+        )
+    }
+}
+
+class SimpleElseCaseExpression<T : ValidType, U : ValidType>(
+    private val case: SimpleCaseClass<T>,
+    private val firstSearchResult: SearchResult<T, out U>,
+    private vararg val additionalSearchResult: SearchResult<T, out U>,
+    private val elseCase: TypeExpression<out U>,
+) : TypeExpression<U> {
+    override fun toDopeQuery(manager: DopeQueryManager): DopeQuery {
+        val caseDopeQuery = case.toDopeQuery(manager)
+        val whenThenDopeQueries = listOf(firstSearchResult, *additionalSearchResult).map {
+            it.searchExpression.toDopeQuery(manager) to it.resultExpression.toDopeQuery(manager)
+        }
+        val elseCaseDopeQuery = elseCase.toDopeQuery(manager)
+        return DopeQuery(
+            queryString = caseDopeQuery.queryString +
+                whenThenDopeQueries.joinToString(" ", " ", " ") { "$WHEN ${it.first.queryString} $THEN ${it.second.queryString}" } +
+                "$ELSE ${elseCaseDopeQuery.queryString} " +
+                END,
+            parameters = caseDopeQuery.parameters + whenThenDopeQueries.fold(emptyMap()) { parameters, query ->
+                parameters + query.first.parameters + query.second.parameters
+            } + elseCaseDopeQuery.parameters,
+        )
+    }
+}
+
+fun <T : ValidType> case(case: TypeExpression<T>) = SimpleCaseClass(case)
+
+fun <T : ValidType, U : ValidType> SimpleCaseClass<T>.whenThen(whenThenExpression: SearchResult<T, U>) =
+    SimpleCaseExpression(this, whenThenExpression)
+
+@JvmName("whenThenWithGeneric")
+fun <T : ValidType, U : ValidType> SimpleCaseExpression<T, U>.whenThen(whenThenExpression: SearchResult<T, U>) =
+    SimpleCaseExpression(case, firstSearchResult, *additionalSearchResult, whenThenExpression)
+
+@JvmName("whenThenWithoutGeneric")
+fun <T : ValidType> SimpleCaseExpression<T, out ValidType>.whenThen(whenThenExpression: SearchResult<T, out ValidType>) =
+    SimpleCaseExpression(case, firstSearchResult, *additionalSearchResult, whenThenExpression)
+
+@JvmName("otherwiseWithGeneric")
+fun <T : ValidType, U : ValidType> SimpleCaseExpression<T, U>.otherwise(elseCase: TypeExpression<U>) =
+    SimpleElseCaseExpression(case, firstSearchResult, *additionalSearchResult, elseCase = elseCase)
+
+@JvmName("otherwiseWithoutGeneric")
+fun <T : ValidType> SimpleCaseExpression<T, out ValidType>.otherwise(elseCase: TypeExpression<out ValidType>) =
+    SimpleElseCaseExpression(case, firstSearchResult, *additionalSearchResult, elseCase = elseCase)
+
+class SearchedCaseExpression<U : ValidType>(
```

Comment:
This class looks like the same as the SimpleCaseExpression but with the BooleanType as the first type T - Cant you introduce inheritance here?

Reply:
done